src/libmeasurement_kit/ooni/dpi_fragment.cpp

```diff
@@ -0,0 +1,373 @@
+void SSLFilter::update() {
+    // If we have data from the network to process, put it the memory BIO for OpenSSL
+    if (!nread->empty()) {
+        int written = BIO_write(rbio, nread->c_str(), nread->length());
+        if (written > 0) {
+            nread->erase(0, written);
+        }
+    }
+
+    // If the application wants to write data out to the network, process it with SSL_write
+    if (!awrite->empty()) {
+        int written = SSL_write(ssl, awrite->c_str(), awrite->length());
+
+        if (!continue_ssl_(written)) {
+            throw std::runtime_error("An SSL error occured.");
+        }
+
+        if (written > 0) {
+            awrite->erase(0, written);
+        }
+    }
+
+    // Read data for the application from the encrypted connection and place it in the string for the app to read
+    while (1) {
+        char *readto = new char[1024];
+        int read = SSL_read(ssl, readto, 1024);
+
+        if (!continue_ssl_(read)) {
+            delete readto;
+            throw std::runtime_error("An SSL error occured.");
+        }
+
+        if (read > 0) {
+            size_t cur_size = aread->length();
+            aread->resize(cur_size + read);
+            std::copy(readto, readto + read, aread->begin() + cur_size);
+        }
+
+        delete readto;
+
+        if (static_cast<size_t>(read) != 1024 || read == 0) break;
+    }
+
+    // Read any data to be written to the network from the memory BIO and copy it to nwrite
+    while (1) {
+        char *readto = new char[1024];
+        int read = BIO_read(wbio, readto, 1024);
+
+        if (read > 0) {
+            size_t cur_size = nwrite->length();
+            nwrite->resize(cur_size + read);
+            std::copy(readto, readto + read, nwrite->begin() + cur_size);
+        }
+
+        delete readto;
+
+        if (static_cast<size_t>(read) != 1024 || read == 0) break;
+    }
+}
+bool SSLFilter::continue_ssl_(int function_return) {
+    int err = SSL_get_error(ssl, function_return);
+    char desc[120];
+
+    if (err != SSL_ERROR_NONE) {
+        ERR_error_string(err, desc);
+        ERR_print_errors_fp(stdout);
+    }
+
+    if (err == SSL_ERROR_NONE || err == SSL_ERROR_WANT_READ) {
+        return true;
+    }
+
+    if (err == SSL_ERROR_SYSCALL) {
+        ERR_print_errors_fp(stderr);
+        perror("syscall error: ");
+        return false;
+    }
+
+    if (err == SSL_ERROR_SSL) {
+        ERR_print_errors_fp(stderr);
+        return false;
+    }
+    return true;
+}
+
+std::string fragmented_https_request(bool do_fragment, bool do_ssl,
+                        SharedPtr<Logger> logger) {
+    std::string hostname { "example.com" };
+    std::string ip { "93.184.216.34" };
+
+    auto nread = std::make_shared<std::string>();
+    auto nwrite = std::make_shared<std::string>();
+    auto aread = std::make_shared<std::string>();
+    auto awrite = std::make_shared<std::string>();
+
+    SSL_CTX* ctx = SSL_CTX_new(SSLv23_method()); //XXX don't ignore errors
```

Comment:
`SSLv23_method()` cannot fail, IIRC. It should return a pointer to a statically allocated `struct`.